cpp/include/cudf/ast/linearizer.hpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/ast/detail/linearizer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/operators.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+
+namespace ast {
+
+/**
+ * @brief Enum of table references.
+ *
+ * This determines which table to use in cases with two tables (e.g. joins).
+ *
+ */
+enum class table_reference {
+  LEFT,   // Column index in the left table
+  RIGHT,  // Column index in the right table
+  OUTPUT  // Column index in the output table
+};
+
+// Forward declaration
+class literal;
+class column_reference;
+class expression;
+
+/**
+ * @brief A literal value used in an abstract syntax tree.
+ *
+ */
+class literal : public detail::node {
+  friend class detail::linearizer;
+
+ public:
+  /**
+   * @brief Construct a new literal object.
+   *
+   * @param value A fixed width scalar device view containing the literal value.
+   */
+  literal(const cudf::detail::fixed_width_scalar_device_view_base value) : value(value) {}
```

Comment:
@jrhemstad I introduced the new base class `detail::fixed_width_scalar_device_view_base` in this PR. I tried to fall in line with the existing public/detail APIs for scalar device views. Classes in the public API like `numeric_scalar_device_view` inherit from the templated class `detail::fixed_width_scalar_device_view<T>`, which in turn inherits from that new `detail::` base class. How should we go about exposing this to the user, so that a `literal` can be constructed without requiring the `detail::` API? It doesn't seem like we should make the templated class public and its parent (my new class) public.

Reply:
I would think that `literal` should be constructible only from `cudf::scalar const&`. The conversion to a `fixed_width_scalar_device_view_base` should happen internally.